packages/firestore/src/local/query_engine.ts

```diff
@@ -15,26 +15,202 @@
  * limitations under the License.
  */
 
-import { SnapshotVersion } from '../core/snapshot_version';
-import { Query } from '../core/query';
-import { DocumentKeySet, DocumentMap } from '../model/collections';
 import { LocalDocumentsView } from './local_documents_view';
 import { PersistenceTransaction } from './persistence';
 import { PersistencePromise } from './persistence_promise';
+import {
+  hasLimitToFirst,
+  hasLimitToLast,
+  LimitType,
+  matchesAllDocuments,
+  newQueryComparator,
+  Query,
+  queryMatches,
+  stringifyQuery
+} from '../core/query';
+import { SnapshotVersion } from '../core/snapshot_version';
+import {
+  DocumentKeySet,
+  DocumentMap,
+  MaybeDocumentMap
+} from '../model/collections';
+import { Document } from '../model/document';
+import { debugAssert } from '../util/assert';
+import { getLogLevel, LogLevel, logDebug } from '../util/log';
+import { SortedSet } from '../util/sorted_set';
 
 /**
- * Represents a query engine capable of performing queries over the local
- * document cache. You must call `setLocalDocumentsView()` before using.
+ * A query engine that takes advantage of the target document mapping in the
+ * QueryCache. Query execution is optimized by only reading the documents that
+ * previously matched a query plus any documents that were edited after the
+ * query was last listened to.
+ *
+ * There are some cases when this optimization is not guaranteed to produce
+ * the same results as full collection scans. In these cases, query
+ * processing falls back to full scans. These cases are:
+ *
+ * - Limit queries where a document that matched the query previously no longer
+ *   matches the query.
+ *
+ * - Limit queries where a document edit may cause the document to sort below
+ *   another document that is in the local cache.
+ *
+ * - Queries that have never been CURRENT or free of Limbo documents.
```

Comment:
ultranit: I think that the "L" in "Limbo" can be lowercase, since "limbo" is not a proper noun.

Reply:
Fixed. I do consider "Limbo" more of a concept, rather than just the state that it refers to. Fixed either way though.